import pandas as pd

data = []
#all genes file, including gene vectors (from motif vectors generated by autoencoder)
all_motif = "motif_50.csv"
#Read the transcription factor gene vectors file
tf_motif = "TF_motif_dim_50_iter_9.txt"

# open file from skip-gram
with open(tf_motif, 'r') as file:
    # 逐行读取
    for line in file:
        # 移除行尾的换行符并分割行
        parts = line.strip().split('\t')
        # 第一个值为行名，其余的值使用空格分隔
        row_name = parts[0]
        values = parts[1].split()
        # 将行名和对应的值添加到列表中
        data.append([row_name] + values)

columns = ['RowName'] + [f'Value{i+1}' for i in range(len(data[0]) - 1)]

TF_vectors = pd.DataFrame(data, columns=columns)

# read tf_list.csv
df_genes = pd.read_csv('tf_list.csv')

# Screening by transcription factor name
filtered_df = TF_vectors[TF_vectors.iloc[:, 0].isin(df_genes['tf'])]

# Save the filter results to a new CSV file
filtered_df.to_csv('TF_motif_50.csv', index=False)

# reads all genes file, including gene vectors (from motif vectors generated by autoencoder)
df1 = pd.read_csv('all_motif', index_col=0)

# Read the transcription factor file
df2 = pd.read_csv('TF_motif_50.csv', index_col=0)

# Identify transcription factors
genes_in_df2 = df2.index

# Replace
df1.loc[genes_in_df2] = df2

# generate gene_motif.csv
df1.to_csv('gene_motif.csv')